From f19ecb10409cc96ee83ac72fdc1cb20b1d36d3b7 Mon Sep 17 00:00:00 2001
From: Mattias Busck <mattias.busck@hostmobility.com>
Date: Wed, 8 Feb 2023 06:45:03 +0100
Subject: [PATCH] Add suspend to optimized tcan4x5x driver

Add sysfile named tcan4x5x_mode that can be set to normal, standby or
sleep. Sleep mode can only wake up on reset pin or CAN traffic and
transfers to standby when this occurs.
---
 drivers/net/can/m_can/tcan4x5x-core.c | 89 +++++++++++++++++++++++++++
 1 file changed, 89 insertions(+)

diff --git a/drivers/net/can/m_can/tcan4x5x-core.c b/drivers/net/can/m_can/tcan4x5x-core.c
index 14894b247654..416040c2da5d 100644
--- a/drivers/net/can/m_can/tcan4x5x-core.c
+++ b/drivers/net/can/m_can/tcan4x5x-core.c
@@ -302,6 +302,87 @@ static struct m_can_ops tcan4x5x_ops = {
 	.clear_interrupts = tcan4x5x_clear_interrupts,
 };
 
+ssize_t tcan4x5x_mode_store(struct device *dev,
+						   struct device_attribute *attr,
+						   const char *buf,
+						   size_t count)
+{
+	int ret;
+	char mode[15];
+	struct tcan4x5x_priv *tcan4x5x = dev_get_drvdata(dev);
+
+	strncpy(mode, buf, count -1);
+
+	if (!strncmp(buf, "normal", count - 1)) {
+		ret = regmap_update_bits(tcan4x5x->regmap, TCAN4X5X_CONFIG,
+				 TCAN4X5X_MODE_SEL_MASK, TCAN4X5X_MODE_NORMAL);
+		if (ret) {
+			dev_err(dev, "Cannot set operation mode to normal\n");
+			return ret;
+		}
+		dev_dbg(dev, "Switched to normal mode\n");
+	} else if (!strncmp(buf, "sleep", count - 1)) {
+		ret = regmap_update_bits(tcan4x5x->regmap, TCAN4X5X_CONFIG,
+				 TCAN4X5X_MODE_SEL_MASK, TCAN4X5X_MODE_SLEEP);
+		if (ret) {
+			dev_err(dev, "Cannot set operation mode to sleep\n");
+			return ret;
+		}
+		dev_dbg(dev, "Switch to normal sleep\n");
+	} else if (!strncmp(buf, "standby", count - 1)) {
+		ret = regmap_update_bits(tcan4x5x->regmap, TCAN4X5X_CONFIG,
+				 TCAN4X5X_MODE_SEL_MASK, TCAN4X5X_MODE_STANDBY);
+		if (ret) {
+			dev_err(dev, "Cannot set operation mode to standby\n");
+			return ret;
+		}
+		dev_dbg(dev, "Switch to normal standby\n");
+	} else {
+		dev_err(dev, "Unsupported mode\n");
+		return -EINVAL;
+	}
+
+	return count;
+}
+
+ssize_t tcan4x5x_mode_show(struct device *dev,
+						  struct device_attribute *attr,
+						  char *buf)
+{
+	unsigned int mode;
+	int ret;
+	struct tcan4x5x_priv *tcan4x5x = dev_get_drvdata(dev);
+
+	ret = regmap_read(tcan4x5x->regmap, TCAN4X5X_CONFIG, &mode);	
+	if (ret < 0) {
+		dev_err(dev, "Can't read mode ctrl register\n");
+		return -EBUSY;
+	}
+
+	switch (mode & TCAN4X5X_MODE_SEL_MASK) {
+	case TCAN4X5X_MODE_STANDBY:
+		return sprintf(buf, "normal sleep [standby]\n");
+	case TCAN4X5X_MODE_SLEEP:
+		return sprintf(buf, "normal [sleep] standby\n");
+	case TCAN4X5X_MODE_NORMAL:
+		return sprintf(buf, "[normal] sleep standby\n");
+	}
+
+	
+	return sprintf(buf, "Unsupported Operation mode\n");
+}
+
+static DEVICE_ATTR(tcan4x5x_mode, S_IWUSR | S_IRUGO, tcan4x5x_mode_show, tcan4x5x_mode_store);
+
+static struct attribute *tcan4x5x_attributes[] = {
+	&dev_attr_tcan4x5x_mode.attr,
+	NULL
+};
+
+static const struct attribute_group tcan4x5x_attr_group = {
+	.attrs = tcan4x5x_attributes,
+};
+
 static int tcan4x5x_can_probe(struct spi_device *spi)
 {
 	struct tcan4x5x_priv *priv;
@@ -341,6 +422,13 @@ static int tcan4x5x_can_probe(struct spi_device *spi)
 		goto out_m_can_class_free_dev;
 	}
 
+	/* Create attributes for sleep control */
+	ret = sysfs_create_group(&(spi->dev.kobj), &tcan4x5x_attr_group);
+	if (ret) {
+		dev_err(&spi->dev, "Can't create sysfs entries\n");
+		goto out_m_can_class_free_dev;
+	}
+
 	priv->spi = spi;
 
 	mcan_class->pm_clock_support = 0;
@@ -392,6 +480,7 @@ static int tcan4x5x_can_remove(struct spi_device *spi)
 {
 	struct tcan4x5x_priv *priv = spi_get_drvdata(spi);
 
+	sysfs_remove_group(&(spi->dev.kobj), &tcan4x5x_attr_group);
 	m_can_class_unregister(&priv->cdev);
 
 	tcan4x5x_power_enable(priv->power, 0);
-- 
2.30.2

